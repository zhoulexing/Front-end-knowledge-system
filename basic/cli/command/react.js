const config = require("../config");
const inquirer = require("inquirer");
const download = require("download-git-repo");
// 改变输出文字的颜色
const chalk = require("chalk");
// ora小图标
const ora = require("ora");
const fs = require("fs");

module.exports = () => {
    inquirer.prompt(config.react.options).then(answers => {  
        cloneProject(answers);
    });
}

// 根据参数克隆项目
function cloneProject({ name, version, redux, css }) {
    const spinner = ora("正在从github下载zlx-build").start();
    download("github:zhoulexing/zlx-build", name, err => {
        if(!err) {
            // 将输出的文字转化为蓝色
            console.info(chalk.blueBright("下载成功"));
            updatePackageName(name);
        } else {
            console.error(err);
        }
        spinner.stop();
    });
}

// // 修改项目名
function updatePackageName(name) {
    fs.readFile(`${ process.cwd() }/${ name }/package.json`, (err, data) => {
        if(err) throw err;
        let _data = JSON.parse(data.toString());
        _data.name = name;
        let str = JSON.stringify(_data, null, 4);
        fs.writeFile(`${ process.cwd() }/${ name }/package.json`, str, err => {
            if(err) throw err;
        });
    });
}