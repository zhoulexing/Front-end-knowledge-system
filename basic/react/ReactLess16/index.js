import ReactDOMTextComponent from "./ReactDOMTextComponent";
import ReactDOMComponent from "./ReactDOMComponent";
import ReactCompositeComponent from "./ReactCompositeComponent";
import ReactElement from "./ReactElement";

// 定义ReactClass类,所有自定义的超级父类
export const ReactClass = function(){
}
// 留给子类去继承覆盖
ReactClass.prototype.render = function(){};

ReactClass.prototype.setState = function(newState) {
    //还记得我们在ReactCompositeComponent里面mount的时候 做了赋值
    //所以这里可以拿到 对应的ReactCompositeComponent的实例_reactInternalInstance
    this._reactInternalInstance.receiveComponent(null, newState);
}

const React = {
    nextReactRootIndex: 0,
    render: function(element, container) {
        // 根据element返回一个component
        const componentInstance = instantiateReactComponent(element);
        // 渲染生成dom结构
        const markup = componentInstance.mountComponent(React.nextReactRootIndex++);
        // 插入到容器
        $(container).html(markup);
        // 触发完成mount的事件
        $(document).trigger("mountReady");
    },
    createElement: (type, config, children) => {
        const props = {};
        const propName;
        config = config || {};
        const key = config.key || null;
        for(propName in config) {
            if(config.hasOwnProperty(propName) && propName !== "key") {
                props[propName] = config[propName];
            }
        }

        const childrenLength = arguments.length - 2;
        if(childrenLength === 1) {
            props.children = Array.isArray(children) ? children : [children];
        } else if(childrenLength > 1) {
            const childArray = Array(childrenLength);
            for (let i = 0; i < childrenLength; i++) {
                childArray[i] = arguments[i + 2];
            }
            props.children = childArray;
        }

        return new ReactElement(type, key, props);
    },
    createClass: (spec) => {
        const Constructor = props => {
            this.props = props;
            this.state = this.getInitialState ? this.getInitialState() : null;
        }
        Constructor.prototype = new ReactClass();
        Constructor.prototype.constructor = Constructor;
        $.extend(Constructor.prototype, spec);
        return Constructor;
    },
};

export function instantiateReactComponent(node) {
    // 如果传进来的node是一个字符串或者一个数值
    if(typeof node === "string" || typeof node === "number") {
        return new ReactDOMTextComponent(node);
    }

    //浏览器基本element,注意基本类型element的type一定是字符串，可以和自定义element时对比
    if(typeof node === "object" && typeof node.type === "string"){
        //注意这里，使用了一种新的component
        return new ReactDOMComponent(node);
    }

    // 自定义的元素节点，类型为构造函数
    if(typeof node === 'object' && typeof node.type === 'function'){
        // 注意这里，使用新的component,专门针对自定义元素
        return new ReactCompositeComponent(node);
    }
}

//用来判定两个element需不需要更新
//这里的key是我们createElement的时候可以选择性的传入的。用来标识这个element，当发现key不同时，我们就可以直接重新渲染，不需要去更新了。
export const _shouldUpdateReactComponent = (prevElement, nextElement) => {
    if (prevElement != null && nextElement != null) {
    var prevType = typeof prevElement;
    var nextType = typeof nextElement;
    if (prevType === 'string' || prevType === 'number') {
      return nextType === 'string' || nextType === 'number';
    } else {
      return nextType === 'object' && prevElement.type === nextElement.type && prevElement.key === nextElement.key;
    }
  }
  return false;
}

export default React;