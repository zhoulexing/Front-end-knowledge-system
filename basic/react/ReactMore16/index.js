function createTextDom(text) {
    return {
        type: "TEXT",
        props: {
            nodeValue: text,
            children: [],
        },
    };
}

function createElement(type, props, ...children) {
    return {
        type,
        props: {
            ...props,
            children: children.map((child) => {
                return typeof child === "object" ? child : createTextDom(child);
            }),
        },
    };
}

// 任务调度
let nextUnitOfWork = null;
let workInProgressRoot = null;
let currentRoot = null;
let deletions = null;
function workLoop(deadline) {
    while (nextUnitOfWork && deadline.timeRemaining() > 1) {
        nextUnitOfWork = performUnitOfWork(nextUnitOfWork);
    }

    if (!nextUnitOfWork && workInProgressRoot) {
        commitRoot();
    }

    requestIdleCallback(workLoop);
}
requestIdleCallback(workLoop);

function commitRoot() {
    deletions.forEach(commitRootImpl);
    commitRootImpl(workInProgressRoot.child);
    currentRoot = workInProgressRoot;
    workInProgressRoot = null;
}

function commitRootImpl(fiber) {
    if (!fiber) {
        return;
    }
    let parentFiber = fiber.return;
    while (!parentFiber.dom) {
        parentFiber = parentFiber.return;
    }
    const parentDom = parentFiber.dom;

    if (fiber.effectTag === "REPLACEMENT" && fiber.dom) {
        parentDom.appendChild(fiber.dom);
    } else if (fiber.effectTag === "DELETION") {
        commitDeletion(fiber, parentDom);
    } else if (fiber.effectTag === "UPDATE" && fiber.dom) {
        updateDom(fiber.dom, fiber.alternate.props, fiber.props);
    }

    commitRootImpl(fiber.child);
    commitRootImpl(fiber.sibling);
}

function createDom(vDom) {
    let dom;
    if (vDom.type === "TEXT") {
        dom = document.createTextNode(vDom.props.nodeValue);
    } else {
        dom = document.createElement(vDom.type);

        if (vDom.props) {
            Object.keys(vDom.props)
                .filter((key) => key !== "children")
                .forEach((item) => {
                    if (item.indexOf("on") === 0) {
                        dom.addEventListener(
                            item.substr(2).toLowerCase(),
                            vDom.props[item],
                            false
                        );
                    } else {
                        dom[item] = vDom.props[item];
                    }
                });
        }
    }

    return dom;
}

function updateDom(dom, prevProps, nextProps) {
    Object.keys(prevProps)
        .filter((name) => name !== "children")
        .filter((name) => !(name in nextProps))
        .forEach((name) => {
            if (name.indexOf("on") === 0) {
                dom.addEventListener(
                    name.substr(2).toLowerCase(),
                    nextProps[name],
                    false
                );
            } else {
                dom[name] = nextProps[name];
            }
        });

    Object.keys(nextProps)
        .filter((name) => name !== "children")
        .forEach((name) => {
            if (name.indexOf("on") === 0) {
                dom.addEventListener(
                    name.substr(2).toLowerCase(),
                    nextProps[name],
                    false
                );
            } else {
                dom[name] = nextProps[name];
            }
        });
}

function commitDeletion(fiber, domParent) {
    if (fiber.dom) {
        domParent.removeChild(fiber.dom);
    } else {
        commitDeletion(fiber.child, domParent);
    }
}

function performUnitOfWork(fiber) {
    const isFunctionComponent = fiber.type instanceof Function;
    if (isFunctionComponent) {
        updateFunctionComponent(fiber);
    } else {
        updateHostComponent(fiber);
    }

    if (fiber.child) {
        return fiber.child;
    }

    let nextFiber = fiber;
    while (nextFiber) {
        if (nextFiber.sibling) {
            return nextFiber.sibling;
        }

        nextFiber = nextFiber.return;
    }
}

function updateHostComponent(fiber) {
    if (!fiber.dom) {
        fiber.dom = createDom(fiber);
    }

    const elements = fiber.props && fiber.props.children;
    reconcileChildren(fiber, elements);
}

function reconcileChildren(workInProgressFiber, elements) {
    let oldFiber =
        workInProgressFiber.alternate && workInProgressFiber.alternate.child;
    let prevSibling = null;
    let index = 0;
    if (elements && elements.length) {
        if (!oldFiber) {
            for (let i = 0; i < elements.length; i++) {
                const element = elements[i];
                const newFiber = buildNewFiber(element, workInProgressFiber);

                if (i === 0) {
                    workInProgressFiber.child = newFiber;
                } else {
                    prevSibling.sibling = newFiber;
                }

                prevSibling = newFiber;
            }
        }

        while (index < elements.length && oldFiber) {
            let element = elements[index];
            let newFiber = null;

            const sameType =
                oldFiber && element && oldFiber.type === element.type;
            if (sameType) {
                newFiber = {
                    type: oldFiber.type,
                    props: element.props,
                    dom: oldFiber.dom,
                    return: workInProgressFiber,
                    alternate: oldFiber,
                    effectTag: "UPDATE",
                };
            } else if (!sameType && element) {
                newFiber = buildNewFiber(element, workInProgressFiber);
            } else if (!sameType && oldFiber) {
                oldFiber.effectTag = "DELETION";
                deletions.push(oldFiber);
            }

            oldFiber = oldFiber.sibling;
            if (index === 0) {
                workInProgressFiber.child = newFiber;
            } else {
                prevSibling.sibling = newFiber;
            }

            prevSibling = newFiber;
            index++;
        }
    }
}

function buildNewFiber(fiber, workInProgressFiber) {
    return {
        type: fiber.type,
        props: fiber.props,
        dom: null,
        return: workInProgressFiber,
        alternate: null,
        effectTag: "REPLACEMENT",
    };
}

function render(vDom, container) {
    workInProgressRoot = {
        dom: container,
        props: {
            children: [vDom],
        },
        alternate: currentRoot,
    };

    deletions = [];

    nextUnitOfWork = workInProgressRoot;
}

let wipFiber = null;
let hookIndex = null;
function useState(init) {
    const oldHook =
        wipFiber.alternate &&
        wipFiber.alternate.hooks &&
        wipFiber.alternate.hooks[hookIndex];
    const hook = {
        state: oldHook ? oldHook.state : init, // state是每个具体的值
    };
    wipFiber.hooks.push(hook);
    hookIndex++;

    const setState = (value) => {
        hook.state = value;

        workInProgressRoot = {
            dom: currentRoot.dom,
            props: currentRoot.props,
            alternate: currentRoot,
        };

        nextUnitOfWork = workInProgressRoot;
        deletions = [];
    };

    return [hook.state, setState];
}

function updateFunctionComponent(fiber) {
    // 支持useState，初始化变量
    wipFiber = fiber;
    hookIndex = 0;
    wipFiber.hooks = []; // hooks用来存储具体的state序列

    // 函数组件的type就是个函数，直接拿来执行可以获得DOM元素
    const children = [fiber.type(fiber.props)];

    reconcileChildren(fiber, children);
}

class Component {
    constructor(props) {
        this.props = props;
    }
}

function transfer(Component) {
    return function (props) {
        const component = new Component(props);
        let [state, setState] = useState(component.state);
        component.props = props;
        component.state = state;
        component.setState = setState;

        return component.render();
    };
}

module.exports = {
    createElement,
    render,
    useState,
    Component,
    transfer
};
