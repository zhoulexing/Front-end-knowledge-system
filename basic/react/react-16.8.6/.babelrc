{
    "presets": [
        "@babel/preset-react",
        [
            "@babel/preset-env",
            {
                "targets": {
                    "browsers": "> 0.25%, not dead"
                }
            }
        ]
    ],
    "plugins": []
}
