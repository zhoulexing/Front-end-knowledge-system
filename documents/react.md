## react 初始化渲染流程

react 渲染主要是分为两个阶段，分别是调和阶段和 commit 阶段。在调和阶段，会指定每个 fiber 的 firstEffect、
lastEffect、nextEffect，在 commit 阶段，就是根据这些 Effect 进行处理。

-   react 渲染的核心逻辑在 ReactDOM 对象上，首先调用 ReactDOM 的 render 方法；
-   第一步会根据 dom 容器即挂载点创建一个 root 对象，root 对象通过\_internalroot 属性能获取 FiberRootNode;
-   FiberRootNode 通过 current 属性能获取 HostRootFiber,HostRootFiber 是顶级的 fiber；
-   每个 fiber 对象上都有三个主要的属性，分别是 stateNode、updateQueue、alternate，
    不同的 fiber，stateNode 指向也会不同，HostRootFiber 的 stateNode 指向 FiberRootNode，
    类 fiber 和函数 fiber 的 stateNode 指向 render 方法返回的 element，原声元素的 stateNode 是原声 dom；
    updateQueue 是用来处理 fiber 更新的，alternate 属性是 fiber 和 workInProgress 相互指向对方的。
-   初始化 fiber 节点后，会调用 scheduleUpdateOnFiber 方法，setState 和 hook 里 useState 的更新装状态方法
    都会调用此方法；
-   scheduleUpdateOnFiber 方法中首先会回去 HostRootFiber,然后调用 performSyncWorkOnRoot，
    performSyncWorkOnRoot 方法中会调用 renderRootSync 方法，当然这中间会进行优先级的计算；
-   renderRootSync 方法会调用 workLoopSync 方法，workLoopSync 方法中，会进入一个大循环，
    从根节点一直遍历调用 performUnitOfWork 方法；
-   performUnitOfWork 方法会调用 beginWork 方法，并获取下一个 Fiber，然后判定下个 Fiber 是否为 null，
    如果是 null 则进入 completeUnitOfWork 方法，不是 null 则继续进行调用 performUnitOfWork 方法。
-   beginWork 方法会根据 workInProgress 的 tag 调用的不同的 mount 方法，类组件会在此过程中调用 render 方法，
    以及 render 方法之前的所有生命周期函数。此外还会获取组件 children 中的元素，根据每个子元素生成对应的 fibre 对象，
    并指定 fiber 之间的关系，最后返回第一个子 fiber 对象。
-   completeUnitOfWork 方法只有在子 fiber 为 null 的时候执行，而子 fiber 为 null 要么就是原声元素、字符串或数字，
    所以这个方法的目的就是生成原声元素、字符串或数字，并将其复制给 stateNode。
-   调和过程结束之后，就看是调用 commitRoot 方法；
-   commitRoot 或 commitRootImpl 方法的核心主要是三个方法，分别是 commitBeforeMutationLifeCycles、
    commitMutationEffects、recursivelyCommitLayoutEffects;
-   commitBeforeMutationLifeCycles 方法会执行组件 getSnapShotBeforeUpdate，当前前提是组件处在更新阶段
    且是 class 组件；
-   commitMutationEffects 方法会将组件的 stateNode 进行拼接或插入到对应的 dom 上，编程最终页面上显示的 dom；
    而且类组件
-   recursivelyCommitLayoutEffects 方法会执行组件渲染完之后的生命周期函数，如果是初始化渲染，
    则会执行 componentDidMount 函数，如果是更新，则会执行 componentDidUpdate 方法。

## react 更新渲染流程

react 只有调用 setState 方法或者调用 hook 的方法来进行更新，而这最终都会调用 scheduleUpdateOnFiber 方法，
后续的流程与初始化流程差不多，只不过在过程中会根据不同的条件进行不同的逻辑处理。如果组件是正在更新中，
则会将 performSyncWorkOnRoot 方法添加到 queueCallack 数组中，待上一次渲染完，在执行 queueCallback 数组
中的函数进行更新。

## react 状态管理库

-   redux
    redux、react-redux、redux-thunk、redux-saga（dva）、redux-promise、redux-observable
-   mobx
    mobx、mobx-react-lite、
-   recoil
    recoil
-   xstate
    xstate、@xstate/react
-   hooks
    use-global-hook、mozz、stated-bean、unstated-next、react-duce、
-   context
    useContext、createContext、

react 渲染流程：

1. reactDOM 在调用 render 方法的时候，首先会调用 legacyRenderSubtreeIntoContainer 方法，
   此方法首先会移除挂载点下的所有子 DOM，然后创建 FiberRootNode 和 HostRootFiber 节点，
   FiberRootNode 的 current 指向 HostRootFiber，HostRootFiber 的 stateNode 指向 FiberRootNode。

2. 然后调用 Reconciler 模块里 updateContainer 方法，此方法会获取当前时间，然后根据 Fiber 获取过期时间，
   最后调用 updateContainerAtExpirationTime 方法。

3. updateContainerAtExpirationTime 方法中会先获取上下文 context,然后赋值给 FiberRootNode 的 context 属性，
   或者 pendingContext 属性，最后调用 scheduleRootUpdate 方法。

4. scheduleRootUpdate 会创建一个 update 对象，然后将 element 组件赋值给 update 对象的 payload 属性；
   紧接着调用 flushPassiveEffects 方法、

问题：

flushPassiveEffects 方法干嘛的？
