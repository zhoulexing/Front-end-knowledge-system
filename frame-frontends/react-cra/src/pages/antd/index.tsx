import React, { Component } from "react";
import { Form, Input } from "antd";

class AntdTest extends Component {
    

    render() {
        return (
            <div>
                <Form>
                    <Form.Item 
                        name="username"
                    >
                        <Input />
                    </Form.Item>
                </Form>
            </div>
        );
    }
}

export default AntdTest;
