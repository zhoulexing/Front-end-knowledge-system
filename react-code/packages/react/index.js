const React = require('./src/React');

module.exports = React.default || React;