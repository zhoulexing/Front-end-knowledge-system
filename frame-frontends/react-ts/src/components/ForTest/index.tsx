import * as React from "react";
import { Button } from "antd";

interface IForTestState {
    count: number,
}

interface IForTestProps {

}


class ForTest extends React.Component<IForTestProps, IForTestState> {
    constructor(props: IForTestProps) {
        super(props);
        this.state = {
            count: 1
        }
    }

    render() {
        return (
            <div>
                <Button onClick={this.getCount}>获取数量</Button>
            </div>
        )
    }

    getCount = () => {
        const { count } = this.state;
        this.setState({
            count: count + 1,
        });
        return count;
    }
}

export default ForTest;