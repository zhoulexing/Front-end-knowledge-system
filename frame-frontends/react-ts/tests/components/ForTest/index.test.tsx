import * as React from "react";
import ForTest from "@/components/ForTest";
import { shallow, mount, render } from "enzyme";
import { Button } from "antd";

describe("test components ForTest", () => {
    beforeAll(() => {
        console.log("global before all");
    });

    afterAll(() => {
        console.log("global after all");
    });

    beforeEach(() => {
        console.log("global before each");
    });

    afterEach(() => {
        console.log("global after each");
    });

    /**
     * shallow：浅渲染，将组件渲染成虚拟DOM对象,只渲染第一层，子组件不会被渲染;
     * render: 静态渲染，将React组件渲染成静态的HTML字符串；
     * mount：完全渲染，它将组件渲染加载成一个真实的DOM节点，用来测试DOM API的交互和组件的生命周期。
     */
    it("shallow get Button length", () => {
        const sWrapper = shallow(<ForTest />);
        const mWrapper = mount(<ForTest />);
        const rWrapper = render(<ForTest />);
        expect(sWrapper.find(Button).length).toBe(1);
        expect(mWrapper.find(Button).length).toBe(1);
        expect(rWrapper.find("button").length).toBe(1);
    });

    /**
     * simulate: 模拟事件
     */
    it("click Button to be done", () => {
        const props = {
            getCount: jest.fn(),
        };
        const sWrapper = shallow(<ForTest {...props} />);
        sWrapper.find(Button).at(0).simulate("click");
        expect(props.getCount).toBeCalled();
    });

    /**
     * UI测试, 生成快照
     */
    it("UI render", () => {
        const props = {
            getCount: jest.fn(),
        };
        const rWrapper = render(<ForTest {...props} />);
        expect(rWrapper).toMatchSnapshot();
    });

    /**
     * 测试组件生命周期
     */
    it("calls componentDidMount", () => {
        const componentDidMountSpy = jest.spyOn(
            ForTest.prototype,
            "componentDidMount"
        );
        shallow(<ForTest />);
        expect(componentDidMountSpy).toHaveBeenCalled();
        // 测试完要进行mockRestore()
        componentDidMountSpy.mockRestore();
    });

    /**
     * 测试组件的内部函数, 用箭头函数来定义方法
     */
    it("calls component getCount", () => {
        const sWrapper = shallow(<ForTest />);
        const spyFunction = jest.spyOn(sWrapper.instance(), "getName");
        sWrapper.instance().getName();
        expect(spyFunction).toHaveBeenCalled();
        spyFunction.mockRestore();
    });

    /**
     * 测试组件的内部函数, 在constructor使用bind来定义方法
     */
    it("calls component getGet", () => {
        const props = {
            getCount: jest.fn(),
        };
        const sWrapper = shallow(<ForTest {...props} />);
        const spyFunction = jest.spyOn(ForTest.prototype, "getAge");
        const Construc = sWrapper.instance();
        Construc.getName();
        expect(spyFunction).toHaveBeenCalled();
        spyFunction.mockRestore();
    });
});
