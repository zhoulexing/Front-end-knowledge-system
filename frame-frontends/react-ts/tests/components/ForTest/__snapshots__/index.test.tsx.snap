// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test components ForTest UI render 1`] = `
<div>
  <button
    class="ant-btn"
    type="button"
  >
    <span>
      获取数量
    </span>
  </button>
</div>
`;
